src/lsp.ts

```diff
@@ -0,0 +1,160 @@
+import * as Admzip from "adm-zip";
+import * as which from "which";
+import * as https from "https";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { LanguageClient } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  config: WorkspaceConfiguration
+  ls: LanguageClient | null = null
+  executableName: string
+  context: ExtensionContext
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName
+    this.context = context
+    this.config = workspace.getConfiguration("mesonbuild");
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
+  }
+
+  abstract startLanguageServer(serverModule: string): void
+
+  abstract supportsSystem(os: string, arch: string): boolean
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
+    this.dispose();
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
+  }
+
+  canDownloadLanguageServer(): boolean {
+    return this.downloadInfo != null && this.findLanguageServer() == null
+  }
+
+  get languageServerPath(): string | null {
+    return this.config.languageServerPath || this.findLanguageServer() || which.sync(this.executableName, { nothrow: true })
+  }
+
+  async setupLanguageServer(): Promise<void> {
+    const lspDir = path.join(this.getExtensionLSPDir(this.context), this.executableName);
+    const downloadInfo = this.downloadInfo
+    fs.rmSync(
+      lspDir,
+      {
+        recursive: true,
+        force: true
+      }
+    )
+    fs.mkdirSync(
+      lspDir,
+      { recursive: true }
+    )
+    const tmpPath = path.join(os.tmpdir(), `lsp-${Date.now()}.zip`);
+    try {
+      this.downloadFile(downloadInfo[0], tmpPath).then(() => {
+        const hashPromise = this.computeFileHash(tmpPath);
```

Comment:
what do you think about moving the hash check to the download function?

Reply:
Not a fan of it, as it won't be composable anymore. One function = one functionality